x/ccv/provider/keeper/proposal.go

```diff
@@ -360,76 +320,21 @@ func (k Keeper) GetPendingConsumerAdditionProp(ctx sdk.Context, spawnTime time.T
 	return prop, true
 }
 
-// BeginBlockInit iterates over the pending consumer addition proposals in order, and creates
-// clients for props in which the spawn time has passed. Executed proposals are deleted.
-//
-// See: https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/methods.md#ccv-pcf-bblock-init1
-// Spec tag:[CCV-PCF-BBLOCK-INIT.1]
+// BeginBlockInit iterates over the initialized consumers chains and creates clients for chains
+// in which the spawn time has passed
 func (k Keeper) BeginBlockInit(ctx sdk.Context) {
-	propsToExecute := k.GetConsumerAdditionPropsToExecute(ctx)
-
-	for i, prop := range propsToExecute {
-		// create consumer client in a cached context to handle errors
+	for _, consumerId := range k.GetInitializedConsumersReadyToLaunch(ctx) {
 		cachedCtx, writeFn := ctx.CacheContext()
-
-		k.SetTopN(cachedCtx, prop.ChainId, prop.Top_N)
-		k.SetValidatorSetCap(cachedCtx, prop.ChainId, prop.ValidatorSetCap)
-		k.SetValidatorsPowerCap(cachedCtx, prop.ChainId, prop.ValidatorsPowerCap)
-		k.SetMinStake(cachedCtx, prop.ChainId, prop.MinStake)
-		k.SetInactiveValidatorsAllowed(cachedCtx, prop.ChainId, prop.AllowInactiveVals)
-
-		for _, address := range prop.Allowlist {
-			consAddr, err := sdk.ConsAddressFromBech32(address)
-			if err != nil {
-				continue
-			}
-
-			k.SetAllowlist(cachedCtx, prop.ChainId, types.NewProviderConsAddress(consAddr))
-		}
-
-		for _, address := range prop.Denylist {
-			consAddr, err := sdk.ConsAddressFromBech32(address)
-			if err != nil {
-				continue
-			}
-
-			k.SetDenylist(cachedCtx, prop.ChainId, types.NewProviderConsAddress(consAddr))
-		}
-
-		err := k.CreateConsumerClient(cachedCtx, &propsToExecute[i])
+		err := k.LaunchConsumer(cachedCtx, consumerId)
 		if err != nil {
-			// drop the proposal
-			ctx.Logger().Info("consumer client could not be created: %w", err)
-			continue
-		}
-
-		consumerGenesis, found := k.GetConsumerGenesis(cachedCtx, prop.ChainId)
-		if !found {
-			// drop the proposal
-			ctx.Logger().Info("consumer genesis could not be created")
+			ctx.Logger().Error("could not launch chain",
```

Comment:
So this will retry a failing consumer chain launch in every single block?

Reply:
because afaict these are not cleaned up or pruned?

Reply:
Yes. Moved `RemoveConsumerFromToBeLaunchedConsumers` higher so that we delete consumer chains that we tried to initialize and failed as follows:
```
record, err := k.GetConsumerInitializationRecord(ctx, consumerId)
if err != nil {
        ...
	continue
}
k.RemoveConsumerFromToBeLaunchedConsumers(ctx, consumerId, record.SpawnTime)

cachedCtx, writeFn := ctx.CacheContext()
...
```

This is also not bullet proof because `GetConsumerInitializationRecord` might return an error.
Let me know if you have any concerns.